pjmedia/src/pjmedia/sdp_neg.c

```diff
@@ -84,6 +120,100 @@ static pj_status_t custom_fmt_match( pj_pool_t *pool,
                                    unsigned a_fmt_idx,
                                    unsigned option);
 
+static pj_status_t validate_pt(pj_pool_t *pool, pjmedia_sdp_neg* neg, 
+                               pjmedia_sdp_session* sdp_sess);
+
+#define START_DYNAMIC_PT 96
+
+/* Swap two codec id */
+static void swap_codec_id(pj_str_t codec_list[], pj_int8_t i, pj_int8_t j)
+{
+    pj_str_t tmp = codec_list[i];
+    codec_list[i] = codec_list[j];
+    codec_list[j] = tmp;
+}
+
+static void sort_codec_id(pj_str_t codec_list[], pj_int8_t codec_cnt)
+{
+    pj_int8_t i;
+
+    for (i = 0; i < codec_cnt; ++i) {
+        pj_int8_t j, max;
+
+        for (max = i, j = i + 1; j < codec_cnt; ++j) {
+            if (pj_stricmp(&codec_list[j], &codec_list[max]) > 0)
+                max = j;
+        }
+
+        if (max != i)
+            swap_codec_id(codec_list, i, max);
+    }
+}
+
+static void get_audio_codec_id(pjmedia_sdp_neg* neg)
+{
+    pjmedia_codec_mgr* mgr = pjmedia_codec_mgr_instance();
+    pj_int8_t max_cnt = PJ_ARRAY_SIZE(neg->codec_list) - neg->codec_list_cnt;
+    pj_int8_t codec_cnt = (mgr->codec_list_cnt > max_cnt)?max_cnt: mgr->codec_list_cnt;
+
+    pj_memcpy(neg->codec_list + neg->codec_list_cnt, mgr->codec_list, 
+              sizeof(pj_str_t) * codec_cnt);
+    neg->codec_list_cnt += codec_cnt;
+}
+
+#if defined(PJMEDIA_HAS_VIDEO) && (PJMEDIA_HAS_VIDEO != 0)
+
+static void get_video_codec_id(pjmedia_sdp_neg* neg)
+{
+    pjmedia_vid_codec_mgr* mgr = pjmedia_vid_codec_mgr_instance();
+    pj_int8_t max_cnt = PJ_ARRAY_SIZE(neg->codec_list) - neg->codec_list_cnt;
+    neg->codec_list_cnt += (pj_int8_t)pjmedia_vid_codec_mgr_get_codec_ids(mgr, max_cnt, 
+                                                   neg->codec_list + neg->codec_list_cnt);
+}
+
+#endif
+
+static void init_codec_map(pj_pool_t *pool, pjmedia_sdp_neg *neg)
+{
+    unsigned i = 0;    
+    unsigned max_med_map = PJ_ARRAY_SIZE(neg->med_pt_map);
+
+    neg->pool = pool;
+    neg->codec_list_cnt = 0;
+
+    /* Get the audio/video codec, and sort them. */
+    get_audio_codec_id(neg);
+#if defined(PJMEDIA_HAS_VIDEO) && (PJMEDIA_HAS_VIDEO != 0)
+    get_video_codec_id(neg);
+#endif
+    sort_codec_id(neg->codec_list, neg->codec_list_cnt);
```

Comment:
separating the audio/video codec strings into different array should be better, so that we don't need to re-sort them.